gaphor/UML/actions/tests/test_activitypropertypage.py

```diff
@@ -1,53 +1,185 @@
+
+    view = model.get_item(0)
+    view.parameter = "in attr: str"
+
+    assert parameter.direction == "in"
+    assert parameter.name == "attr"
+    assert parameter.typeValue == "str"
+
+
+def test_activity_parameter_node_add_new_parameter(element_factory):
+    activity = element_factory.create(UML.Activity)
+    model = activity_parameter_node_model(activity)
 
-    assert model[iter][-1] is activity_item.subject.node[0]
+    view = model.get_item(0)
+    view.parameter = "in attr: str"
+    parameter = activity.node[0].parameter
 
+    assert view.node in activity.node
+    assert parameter.direction == "in"
+    assert parameter.name == "attr"
+    assert parameter.typeValue == "str"
 
-def test_activity_parameter_node_reorder(create, element_factory):
+
+def test_update_model_when_new_parameter_added(create, element_factory):
     activity_item = create(ActivityItem, UML.Activity)
 
-    def activity_parameter_node(name):
-        node = element_factory.create(UML.ActivityParameterNode)
-        node.parameter = element_factory.create(UML.Parameter)
-        node.name = name
-        activity_item.subject.node = node
-        return node
+    property_page = ActivityItemPage(activity_item)
+    property_page.construct()
+
+    activity_item.subject.node = activity_parameter_node(element_factory, "one")
 
-    node1 = activity_parameter_node("param1")
-    node2 = activity_parameter_node("param2")
-    node3 = activity_parameter_node("param3")
+    assert property_page.model.get_n_items() == 2  # one + empty row
 
-    list_store = ActivityParameters(activity_item)
 
-    new_order = [node3, node1, node2]
-    list_store.sync_model(new_order)
+def test_update_model_when_new_parameter_added_via_list_model(create, element_factory):
+    activity_item = create(ActivityItem, UML.Activity)
+
+    property_page = ActivityItemPage(activity_item)
+    property_page.construct()
+
+    new_view = property_page.model.get_item(0)
+    new_view.parameter = "new"
 
-    assert activity_item.subject.node[0] is node3
-    assert activity_item.subject.node[1] is node1
-    assert activity_item.subject.node[2] is node2
+    assert property_page.model.get_n_items() == 2  # one + empty row
+    assert property_page.model.get_item(0).node is new_view.node
+    assert property_page.model.get_item(1).node is None
 
 
-def test_activity_page_add_attribute(create):
+def test_update_model_with_single_node_when_parameter_deleted(create, element_factory):
     activity_item = create(ActivityItem, UML.Activity)
+    activity_item.subject.node = node = activity_parameter_node(element_factory, "one")
+
     property_page = ActivityItemPage(activity_item)
+    property_page.construct()
+
+    del activity_item.subject.node[node]
+
+    assert not activity_item.subject.node
+    assert property_page.model.get_n_items() == 1
+    assert property_page.model.get_item(0).node is None
+
+
+def test_update_model_with_multiple_noes_when_parameter_deleted(
```

Comment:
Super nit, but should this be nodes instead of noes?

Reply:
fixed